File: packages/host4_flutter_aivoice/ios/AiVoice_副本/Tools/SubtitleMsgData.m
//
//  SubtitleMsgData.m
//  GameMacro
//
//  Created by 敬攀 on 2025/12/10.
//

#import "SubtitleMsgData.h"

@implementation SubtitleMsgData

+ (NSString *)unPackData:(NSData *)message{
    NSString *subtitles = subvUnpack(message);
    
    return subtitles;
}

+(NSMutableArray *)parseMsg:(NSString *)msg{
    return subvParseData(msg);
}

// 大端序转换
uint32_t swapUInt32(uint32_t value) {
    return ((value & 0x000000FF) << 24) |
           ((value & 0x0000FF00) << 8) |
           ((value & 0x00FF0000) >> 8) |
           ((value & 0xFF000000) >> 24);
}

//拆包校验
NSString *subvUnpack(NSData *data) {
    const int kSubtitleHeaderSize = 8;
    
    NSUInteger size = data.length;
    
    if (size < kSubtitleHeaderSize) {
        return nil;
    }
    
    const uint8_t *message = data.bytes;
    // Check magic number "subv"
    uint32_t magic = (message[0] << 24) | (message[1] << 16) | (message[2] << 8) | message[3];
    
    if (magic != 0x73756276) {
        return nil;
    }
    
    // Get length
    uint32_t length = (message[4] << 24) | (message[5] << 16) | (message[6] << 8) | message[7];
    
    if (size - kSubtitleHeaderSize != length) {
        return nil;
    }
    
    // Get subtitles
    NSString *subtitles = nil;
    
    if (length > 0) {
        subtitles = [[NSString alloc] initWithBytes:message + kSubtitleHeaderSize length:length encoding:NSUTF8StringEncoding];
    } else {
        subtitles = @"";
    }
    
    return subtitles;
}

//解析
NSMutableArray *subvParseData(NSString *msg) {
    NSError *error = nil;
    
    NSDictionary *json_data = [NSJSONSerialization JSONObjectWithData:[msg dataUsingEncoding:NSUTF8StringEncoding] options:0 error:&error];
    
    if (error) {
        NSLog(@"JSON Parse Error: %@", error);
        return nil;
    }
    
    NSMutableArray<SubtitleMsgData *> *subtitles = [NSMutableArray array];
    
    for (NSDictionary *item in json_data[@"data"]) {
        SubtitleMsgData *subData = [[SubtitleMsgData alloc] init];
        
        subData.definite = [item[@"definite"] boolValue];
        subData.language = item[@"language"];
        subData.paragraph = [item[@"paragraph"] boolValue];
        subData.sequence = [item[@"sequence"] integerValue];
        subData.text = item[@"text"];
        subData.userId = item[@"userId"];
        subData.roundId = [item[@"roundId"] intValue];
        
        if ([subData.userId isEqualToString:AIVoiceShared.userId]){
            subData.msgType = 0;
        }else{
            subData.msgType = 1;
        }
        
        [subtitles addObject:subData];
    }
    
    return subtitles;
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/Models/SubtitleMsgData.swift
import Foundation

/// subv 消息 - 字幕数据
class SubtitleMsgData {
  var text: String      /// 可变，用于特殊字符替换
  let language: String
  let userId: String
  let sequence: Int
  let roundId: Int
  let definite: Bool    /// 完整分句
  let paragraph: Bool   /// 段落结束
  var msgType: Int      /// 0=用户消息, 1=机器人消息

  init(text: String, language: String, userId: String,
       sequence: Int, roundId: Int, definite: Bool,
       paragraph: Bool, msgType: Int) {
    self.text = text
    self.language = language
    self.userId = userId
    self.sequence = sequence
    self.roundId = roundId
    self.definite = definite
    self.paragraph = paragraph
    self.msgType = msgType
  }

  /// 拆包校验：从二进制数据中提取 subv JSON 字符串
  /// magic number = 0x73756276 ("subv")
  static func unpack(from data: Data) -> String? {
    let headerSize = 8
    guard data.count >= headerSize else { return nil }

    let magic = data.withUnsafeBytes { ptr in
      ptr.loadUnaligned(fromByteOffset: 0, as: UInt32.self).bigEndian
    }
    guard magic == 0x7375_6276 else { return nil }

    let length = data.withUnsafeBytes { ptr in
      ptr.loadUnaligned(fromByteOffset: 4, as: UInt32.self).bigEndian
    }
    guard data.count - headerSize == length else { return nil }

    if length == 0 { return "" }
    return String(data: data.subdata(in: headerSize..<data.count), encoding: .utf8)
  }

  /// 解析 JSON 字符串为 [SubtitleMsgData]
  /// JSON 格式: {"data": [{text, language, userId, sequence, roundId, definite, paragraph}, ...]}
  static func parse(json: String, currentUserId: String) -> [SubtitleMsgData] {
    guard let jsonData = json.data(using: .utf8),
          let dict = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
          let items = dict["data"] as? [[String: Any]]
    else { return [] }

    return items.map { item in
      let userId = item["userId"] as? String ?? ""
      let msgType = (userId == currentUserId) ? 0 : 1

      return SubtitleMsgData(
        text: item["text"] as? String ?? "",
        language: item["language"] as? String ?? "",
        userId: userId,
        sequence: item["sequence"] as? Int ?? 0,
        roundId: item["roundId"] as? Int ?? 0,
        definite: item["definite"] as? Bool ?? false,
        paragraph: item["paragraph"] as? Bool ?? false,
        msgType: msgType
      )
    }
  }
}
